### ISSUE TITLE
UTF8 - Nightmare

### ISSUE THREAD

**Issue by celogeek** [elasticsearch/elasticsearch-perl#57]
Hi,

I'm trying to use Search::Elasticsearch with utf8 text.

```
perl -MDDP -Mutf8 -MSearch::Elasticsearch -E 'my $els = Search::Elasticsearch->new(serializer => "JSON::XS"); my $d = $els->index(qw(index test type test id test body), { test => "héhé" }); say p($d)'
```

When I check the token, I've got properly the word "héhé" searchable. So storing is perfect.

When I want to retrieve the "test", it return "héhé" but in latin1 ! : h<E9>h<E9>

```
perl -MDDP -Mutf8 -MSearch::Elasticsearch -E 'my $els = Search::Elasticsearch->new(serializer => "JSON::XS"); my $d = $els->get(qw(index test type test id test)); say p($d)' 
```

How can I properly set utf8 content to elasticsearch, and retrieve it in the exact same way I have put it ?

It is a problem with the serializer ?

**Comment by celogeek** [comment]
The JSON encode/decode is not obvious.

Here a bitcode :

```
perl -MDDP -Mutf8 -MJSON -E 'my $x = JSON->new->utf8->encode({test => "héhé"}); say $x; p(JSON->new->decode($x))'
{"test":"héhé"}
\ {
    test   "héhé"
}
```

So, to encode utf8 string, we need to activate the "utf8" from JSON, and to decode a string that has been encoded like this, we need to avoid using the "utf8" flag.

I try to create a serializer that works

**Comment by celogeek** [comment]
Oh,

I have solve my problem !

I read my file using the "raw". The file contain "utf8" string.

So when I encode for ElasticSearch, I need to ask JSON to not encode my string.
When I retrieve the information from ElasticSearch, I want the utf8 version, also here I ask to not decode again.

So I have create a serializer like this :

```
package Search::Elasticsearch::Serializer::CGJSON;
use Moo;
use JSON::MaybeXS 1.002002 ();
 
has 'JSON' => ( is => 'ro', default => sub { JSON::MaybeXS->new } );
 
with 'Search::Elasticsearch::Role::Serializer::JSON';
use namespace::clean;
 
1;
```

And now I put and get utf8 string properly.

**Comment by clintongormley** [comment]
Hi @celogeek 

OK - you're doing this all wrong :)  Search::Elasticsearch handles encoding just fine.  The problem you're having is with your terminal (and this is a lesson that every Perl user has to learn).

When moving data to or from a Perl program you need to:

* decode incoming (eg UTF8) bytes into Perl strings 
* encode outgoing Perl strings into (eg UTF8) bytes

I'll step through your first example:

    perl -MDDP -Mutf8 -MSearch::Elasticsearch -E 'my $els = Search::Elasticsearch->new(serializer => "JSON::XS"); my $d = $els->index(qw(index test type test id test body), { test => "héhé" }); say p($d)'

Here, the input is the string `héhé`, which appears within your Perl source code.  These are bytes, encoded by your editor or your terminal into some encoding.  These days, this encoding is almost always UTF8.  So to convert that into a Perl string, you need to decode these UTF8 bytes.  You've got `-Mutf8` which does exactly this.  It assumes that all embedded strings in the Perl program are encoded in UTF8 and decodes them.  

You then pass that Perl string to Search::Elasticsearch (and JSON::XS) which encodes the JSON request (including the Perl string) back into UTF8 bytes before sending to Elasticsearch.

Well and good!  The next step is where things go wrong:

    perl -MDDP -Mutf8 -MSearch::Elasticsearch -E 'my $els = Search::Elasticsearch->new(serializer => "JSON::XS"); my $d = $els->get(qw(index test type test id test)); say p($d)' 

Here you retrieve the (UTF8 encoded) bytes from Elasticsearch, then JSON::XS decodes these bytes into a hash containing Perl strings.  Now you try to output the value to STDOUT.  Perl simply prints out the bytes that are used internally to represent the Perl string, which in this case is likely to be ISO-8859-1.

You need to explicitly tell Perl that it should encode bytes going to STDOUT as UTF8, as that is what your terminal understands. One way of doing this is with the `-C` flag (see http://perldoc.perl.org/perlrun.html#Command-Switches), eg:

    perl -CS -MDDP -Mutf8 -MSearch::Elasticsearch -E 'my $els = Search::Elasticsearch->new(serializer => "JSON::XS"); my $d = $els->get(qw(index test type test id test)); say p($d)' 

Or use `binmode`:

    perl -MDDP -Mutf8 -MSearch::Elasticsearch -E 'binmode(STDOUT,":utf8"); my $els = Search::Elasticsearch->new(serializer => "JSON::XS"); my $d = $els->get(qw(index test type test id test)); say p($d)' 

Alternatively, you can use the `Encode` module to work with individual strings:

    perl  -MDDP -Mutf8 -MSearch::Elasticsearch -E 'use Encode qw(encode_utf8); my $els = Search::Elasticsearch->new(serializer => "JSON::XS"); my $d = $els->get(qw(index test type test id test)); $d->{_source}{test} = encode_utf8($d->{_source}{test}); say p($d)' 

In summary: decode any incoming UTF8 bytes into Perl strings, and encode any outgoing Perl strings into UTF8 bytes, and all will work perfectly.

**Comment by celogeek** [comment]
Yeah, I finally understand the process.

Let me explain my process :

I read an 100% sure UTF-8 file : '<:utf8'
So I got a perl ref { tile => 'perl string', description => 'perl string' ... }

Then I save it to ELS
Search::Elasticsearch take the perl ref, convert it into valid utf8 json string.

Then later I retrieve from ELS and got a ref containing perl string.

In order to display, I need to reencode the all ref into a ref of utf8 string, and send it to the browser.

In short :

utf8 file => decode => ref of perl string => json reencode utf8 => save to els => retrieve a ref of perl string => reencode into a ref of utf8 => display to the browser.

I think they is a lot of extra encode / decode here, and we can save a lot of useless process.

So I have create a simple CGJSON which is the same as your JSON module, but without the utf8 flag on.

The process in short became :

read utf8 file in '<:raw' mode => ref of utf8 string => json encode => save to els => retrieve a ref of utf8 string => display

So no more encode / decode. I read a utf8, I save like this without touching anything, I read it without change and I output it directly.

I wonder if it is possible, like you check when the body is directly a string, to do a JSON->encode then check if the contain is not already an utf8 string (is_utf8), if not encode it, then send.
And in the decode process, you receive a valid json in utf8, so just decode without any more decoding process.

May be I'm alone in that case, but I don't like extract encode/decode when from source to destination I'm sure I got utf8 all along.

**Comment by renormalist** [comment]
@celogeek - You are not alone. :blush:
